### PR TITLE
feat: add option for stacked line layers

### PR DESCRIPTION
**Pull request by TCL735**
Closes https://github.com/influxdata/giraffe/issues/43

### CODE REVIEW

**Comment by TCL735** [comment]
Example of stacked line layer. Notice also the tooltip has an additional column named "cumulative" which reflects the value of the hovered points, while the "_value" is the original single values of the points.

Screen Shot 2019-11-25 at 10 37 04 AM

**Comment by TCL735** [comment]
With shading

Screen Shot 2019-11-25 at 10 42 51 AM

File: giraffe/src/transforms/line.ts [hunk 1]
```diff
@@ -1,13 +1,30 @@
...
+  domainLabel: string
+): number => {
+  if (groupID < 1) {
+    return 0
+  }
+  const currentGroupLength = lineData[groupID][domainLabel].length
+  const previousGroupLength = lineData[groupID - 1][domainLabel].length
+  if (previousGroupLength > currentGroupLength) {
+    return lineData[groupID - 1][domainLabel][currentGroupLength]
+  }
```
Comment (hoorayimhelping) [inline giraffe/src/transforms/line.ts]: this block has a couple of potential undefined accesses (`cannot access property 'length' of null`). not sure how certain we can be that this data exists, but if we want to be more guarded we could do something like:
```ts
const currentGroup = lineData[groupID]
const previousGroup = lineData[groupID - 1]

if (!currentGroup || !previousGroup) {
  return 0
}

const currentDomainLabel = currentGroup[domainLabel]
const previousDomainLabel = previousGroup[domainLabel]

if (currentDomainLabel && previousDomainLabel) {
  if (previousDomainLabel.length > currentDomainLabel.length) {
    return previousDomainLabel[currentDomainLabel.length]
  }
}

return 0
```

alternatively, you could be kinda lazy and yolo it (not a fan of this):

```ts
try {
  const currentGroupLength = lineData[groupID][domainLabel].length
  const previousGroupLength = lineData[groupID - 1][domainLabel].length
  if (previousGroupLength > currentGroupLength) {
    return lineData[groupID - 1][domainLabel][currentGroupLength]
  }
} catch () {
  // chomp
  return 0
}
```

File: giraffe/src/utils/lineData.ts [hunk 2]
```diff
@@ -51,3 +51,15 @@ export const simplifyLineData = (
...
+  lineData: LineData,
+  domainLabel: string
+): Array<number> => {
+  let result = []
+  const numberOfLines = Object.keys(lineData).length
+  for (let index = 0; index < numberOfLines; index += 1) {
+    result = result.concat(lineData[index][domainLabel])
+  }
+  return result
+}
```
Comment (hoorayimhelping) [inline giraffe/src/utils/lineData.ts]: would love to see even a simple test for this!

File: giraffe/src/utils/tooltip.ts [hunk 3]
```diff
@@ -38,6 +47,25 @@ const getTooltipGroupColumns = (
...
+  const yColData = getDomainDataFromLines(lineData, 'ys')
+  const yFormatter = getValueFormatter(yColKey)
+  return {
+    key: yColKey,
+    name: 'cumulative',
+    type: table.getColumnType(yColKey),
+    colors: rowColors,
+    values: rowIndices.map(i => yFormatter(yColData[i])),
+  }
+}
```
Comment (hoorayimhelping) [inline giraffe/src/utils/tooltip.ts]: some jest tests for this would be fantastic

File: giraffe/src/transforms/line.ts [hunk 4]
```diff
@@ -1,13 +1,30 @@
-import {Table, LineLayerSpec, LineData} from '../types'
+import {Table, LineLayerSpec, LineData, LinePosition} from '../types'
 import {FILL} from '../constants/columnKeys'
 import {createGroupIDColumn, getNominalColorScale} from './'
 
+const getPrevCumulativeDomainValue = (
+  lineData: LineData,
+  groupID: number,
+  domainLabel: string
```
Comment (ebb-tide) [inline giraffe/src/transforms/line.ts]: I'm surprised this was not typed previously. Would you be willing to try typing 'xs' and 'ys' as an enum?

File: giraffe/src/types/index.ts [hunk 5]
```diff
@@ -82,12 +82,15 @@ export type LineInterpolation =
...
 
+export type LinePosition = 'overlaid' | 'stacked'
+
 export interface LineLayerConfig {
   type: 'line'
   x: string
   y: string
   fill?: string[]
   colors?: string[]
+  position?: LinePosition
```
Comment (ebb-tide) [inline giraffe/src/types/index.ts]: do you think it would make sense to provide a default here?

File: giraffe/src/transforms/line.ts (hunk 4 above)
Comment (TCL735) [reply giraffe/src/transforms/line.ts]: Sure thing. Will add that to the next push.

File: giraffe/src/types/index.ts (hunk 5 above)
Comment (TCL735) [reply giraffe/src/types/index.ts]: Yes, makes perfect sense. I left it out by mistake. Will add that in with the next push.

File: giraffe/src/transforms/line.ts (hunk 1 above)
Comment (TCL735) [reply giraffe/src/transforms/line.ts]: Yes, I thought a while on this block. `return 0` happening twice did not feel ideal to me and i would like to come up with something more elegant here. I will look into your suggestion as I write the tests.

File: giraffe/src/utils/lineData.ts (hunk 2 above)
Comment (TCL735) [reply giraffe/src/utils/lineData.ts]: Will add the tests with the next push.

File: giraffe/src/utils/tooltip.ts (hunk 3 above)
Comment (TCL735) [reply giraffe/src/utils/tooltip.ts]: Will add the tests with the next push.

**Comment by TCL735** [comment]
Aside from the reviews, I noticed upon further testing that sometimes a line's data will not be in the tooltip when the _value is 0. This happens intermittently, as sometimes the tooltip will correctly display the entire line's data. Will investigate this and report back.

**Comment by TCL735** [comment]
The issue with the tooltip missing a line when _value is 0 is due to domain stretching. When the lines are stacked, the Y domain is stretched (as compared to overlaid position) from adding all of the values together. This causes the hover points to be estimated when the number of time ticks are too few. We'll look into adding a possible solution here. If not possible, then a separate issue will be opened.

**Comment by TCL735** [comment]
> Aside from the reviews, I noticed upon further testing that sometimes a line's data will not be in the tooltip when the _value is 0. This happens intermittently, as sometimes the tooltip will correctly display the entire line's data when _value is 0. Will investigate this and report back.

This is a data issue. The sample test data does not have enough records to allow the hover points to be found with high precision when a line has a _value 0 (and thus causing two line layers to intersect when stacked). Increasing the number of records in the sample data makes this issue un-reproduceable. Intersecting lines will have their own hover points.

In real applications, this issue is moot as the data volume is expected to be much higher than the sample test data in storybook.

File: giraffe/src/transforms/line.ts [hunk 6]
```diff
@@ -28,13 +51,17 @@ export const lineTransform = (
   for (let i = 0; i < table.length; i++) {
```
Comment (hoorayimhelping) [inline giraffe/src/transforms/line.ts]: not a blocker, just a thought for the future: might consider converting this to `forEach` function if at all possible and unless there's a strong reason to use a `for` loop - manually iterating is sooooooooo 1989

**Pull request merged**